routing/index_graph_serializer.hpp

```diff
@@ -0,0 +1,381 @@
+
+    JointIdConverter jointIdConverter(header.GetNumJoints());
+
+    for (uint32_t i = 0; i < header.GetNumSections(); ++i)
+    {
+      Section const & section = header.GetSection(i);
+      VehicleMask const mask = section.GetMask();
+
+      if (!(mask & requiredMask))
+      {
+        src.Skip(section.GetSize());
+        continue;
+      }
+
+      JointIdDecoder jointIdDecoder(section.GetBeginJointId());
+      BitReader<Source> reader(src);
+      uint64_t const expectedEndPos = src.Pos() + section.GetSize();
+
+      // -1 for uint32_t is some confusing, but it allows process first iteration in the common way.
+      // It works because uint32_t is modular ring type.
+      uint32_t featureId = -1;
+      for (uint32_t i = 0; i < section.GetNumRoads(); ++i)
+      {
+        uint32_t const featureDelta = Read32(reader);
+        featureId += featureDelta;
+
+        uint32_t const pointCap = Read32(reader);
+        if (pointCap < 1)
+          MYTHROW(RoutingException, ("Invalid pointCap =", pointCap, ", featureId =", featureId));
+
+        uint32_t const maxPointId = pointCap - 1;
+
+        RoadJointIds & roadJoints = graph.InitRoad(featureId, maxPointId);
+
+        for (uint32_t pointId = -1; pointId != maxPointId;)
+        {
+          uint32_t const pointDelta = Read32(reader);
+          pointId += pointDelta;
+          if (pointId > maxPointId)
+          {
+            MYTHROW(RoutingException, ("Invalid pointId =", pointId, ", maxPointId =", maxPointId,
+                                       ", pointDelta =", pointDelta, ", featureId =", featureId));
+          }
+
+          Joint::Id const jointIdInFile = jointIdDecoder.Read(reader);
+          if (jointIdInFile >= section.GetEndJointId())
+          {
+            MYTHROW(RoutingException,
+                    ("Invalid jointId =", jointIdInFile, ", end =", section.GetEndJointId(),
+                     ", mask =", mask, ", pointId =", pointId, ", featureId =", featureId));
+          }
+
+          Joint::Id const jointId = jointIdConverter.Convert(jointIdInFile);
+          // TODO: filter redundant joints
+          roadJoints.AddJoint(pointId, jointId);
+        }
+      }
+
+      if (jointIdDecoder.GetCount() != section.GetEndJointId())
+      {
+        MYTHROW(RoutingException, ("Invalid decoder count =", jointIdDecoder.GetCount(),
+                                   ", expected =", section.GetEndJointId(), ", mask =", mask));
+      }
+
+      if (src.Pos() != expectedEndPos)
+      {
+        MYTHROW(RoutingException,
+                ("Wrong position", src.Pos(), "after decoding section", mask, "expected",
+                 expectedEndPos, "section size =", section.GetSize()));
+      }
+    }
+
+    graph.Build(jointIdConverter.GetCount());
+  }
+
+private:
+  static uint8_t constexpr kVersion = 0;
+
+  class Section final
+  {
+  public:
+    Section() = default;
+
+    Section(VehicleMask mask, uint64_t size, uint32_t numRoads, Joint::Id beginJointId,
```

Comment:
In general, it would be nice if parameter list will be the same as in Serialize().